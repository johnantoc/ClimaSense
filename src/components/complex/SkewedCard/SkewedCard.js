/**
 * Skewed Card Component.
 */
import React, { useState, useEffect } from "react";
import { View, Image, ActivityIndicator } from "react-native";
import { Entypo } from "@expo/vector-icons";
import { useSelector, useDispatch } from "react-redux";
import * as Location from "expo-location";

import RichText from "../../basic/RichText/RichText";
import SkewedCardStyles from "./SkewedCard.style";
import { getLocation, getWeatherData } from "../../../store/actions";

const SkewedCard = () => {
  const dispatch = useDispatch();
  const [locDenied, setLocDenied] = useState(false);
  const {
    weather,
    settings: { tempUnitF },
    location,
  } = useSelector((state) => {
    return {
      weather: state.weatherData,
      settings: state.settings,
      location: state.location,
    };
  });

  useEffect(() => {
    (async () => {
      const { status } = await Location.requestPermissionsAsync();
      if (status === "granted") {
        const locationData = await Location.getCurrentPositionAsync({
          accuracy: 4,
        });
        dispatch(getLocation(locationData));
      } else {
        setLocDenied(true);
      }
    })();
  }, []);

  useEffect(() => {
    if (location && location.selectedLoc !== null) {
      dispatch(getWeatherData(location));
    }
  }, [location]);

  const temp = weather.currentData
    ? tempUnitF
      ? weather.currentData.tempF
      : weather.currentData.tempC
    : "-";

  return !locDenied || weather.currentData ? (
    <View style={[SkewedCardStyles.outerContainer]}>
      <ActivityIndicator
        size="large"
        color="#FFFFFF"
        style={SkewedCardStyles.loader}
        animating={!weather.currentData}
      />
      <View style={[SkewedCardStyles.container]}>
        <View style={[SkewedCardStyles.mainArea]}>
          <View style={[SkewedCardStyles.dayPlaceStrip]}>
            <View style={[SkewedCardStyles.dayContainer]}>
              <View style={[SkewedCardStyles.textContainer]}>
                <RichText text="Today" styles={SkewedCardStyles.today} />
              </View>
              <View style={[SkewedCardStyles.textContainer]}>
                <RichText
                  text={
                    weather.currentData ? weather.currentData.dateStr : "--"
                  }
                  styles={SkewedCardStyles.date}
                />
              </View>
            </View>
            <View style={[SkewedCardStyles.placeContainer]}>
              <Entypo name="location-pin" size={24} color="#FFFFFF" />
              <RichText
                text={
                  location
                    ? location.selectedLoc
                      ? location.selectedLoc
                      : location.city
                      ? location.city
                      : location.name
                    : "--"
                }
                styles={SkewedCardStyles.place}
              />
            </View>
          </View>
          <View style={[SkewedCardStyles.temperatureContainer]}>
            <RichText text={temp} styles={SkewedCardStyles.temperature} />
            <RichText text="o" styles={SkewedCardStyles.degree} />
            <RichText
              text={tempUnitF ? "F" : "C"}
              styles={SkewedCardStyles.temperatureUnit}
            />
          </View>
        </View>
      </View>
    </View>
  ) : (
    <View style={[SkewedCardStyles.errorContainer]}>
      <ActivityIndicator
        size="large"
        color="#FFFFFF"
        style={SkewedCardStyles.loader}
        animating={location && location.selectedLoc !== null}
      />
      <Image
        style={SkewedCardStyles.logo}
        source={require("../../../assets/logowhite.png")}
      />
      <RichText
        text={"Please enable permission to access the location"}
        styles={SkewedCardStyles.errorTxt}
      />
    </View>
  );
};

export default SkewedCard;
