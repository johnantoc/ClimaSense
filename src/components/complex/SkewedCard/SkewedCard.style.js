import { StyleSheet } from "react-native";

const SkewedCardStyles = StyleSheet.create({
  outerContainer: {
    flex: 0.4,
    backgroundColor: "transparent",
  },
  container: {
    flex: 1,
    backgroundColor: "transparent",
  },
  text: {
    color: "#FFFFFF",
  },
  mainArea: {
    flex: 1,
    opacity: 1,
  },
  dayPlaceStrip: {
    flex: 0.4,
    paddingHorizontal: 15,
    flexDirection: "row",
  },
  dayContainer: {
    flex: 1,
    justifyContent: "center",
    alignItems: "flex-start",
  },
  textContainer: {
    marginBottom: 4,
  },
  today: {
    color: "#FFFFFF",
    fontSize: 18,
    fontFamily: "roboto-condensed-light",
  },
  date: {
    color: "#FFFFFF",
    fontSize: 16,
    fontFamily: "roboto-condensed-light",
  },
  placeContainer: {
    flex: 1,
    paddingTop: 10,
    flexDirection: "row",
    justifyContent: "flex-end",
  },
  place: {
    color: "#FFFFFF",
    fontSize: 20,
    marginLeft: 8,
    textTransform: "capitalize",
    fontFamily: "roboto-condensed-light",
  },
  temperatureContainer: {
    flex: 0.6,
    flexDirection: "row",
    paddingLeft: 15,
    paddingRight: 15,
  },
  temperature: {
    color: "#FFFFFF",
    fontSize: 70,
    marginRight: 10,
    fontFamily: "roboto-condensed-light",
  },
  degree: {
    color: "#FFFFFF",
    fontSize: 30,
    fontFamily: "roboto-condensed-light",
  },
  temperatureUnit: {
    color: "#FFFFFF",
    fontSize: 50,
    fontFamily: "roboto-condensed-light",
  },
  loader: {
    position: "absolute",
    top: "20%",
    left: "46%",
  },
  errorContainer: {
    flex: 1,
    justifyContent: "center",
    alignItems: "center",
  },
  errorTxt: {
    color: "#FFFFFF",
    fontSize: 16,
    textTransform: "capitalize",
    fontFamily: "roboto-condensed-light",
  },
  logo: {
    width: 120,
    height: 85,
  },
});

export default SkewedCardStyles;
