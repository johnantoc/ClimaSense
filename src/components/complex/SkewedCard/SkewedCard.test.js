import React from 'react';
import SkewedCard from './SkewedCard';

import renderer from 'react-test-renderer';

test('SkewedCard renders correctly', () => {
  const tree = renderer.create(<SkewedCard />).toJSON();
  expect(tree).toMatchSnapshot();
});
