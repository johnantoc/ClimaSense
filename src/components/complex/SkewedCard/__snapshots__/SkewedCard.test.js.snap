// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SkewedCard renders correctly 1`] = `
<View
  style={
    Array [
      Object {
        "backgroundColor": "#FFFFFF",
        "flex": 0.4,
        "height": 100,
      },
    ]
  }
>
  <View
    style={
      Array [
        Object {
          "backgroundColor": "#000119",
          "borderBottomRightRadius": 80,
          "flex": 1,
        },
      ]
    }
  />
</View>
`;
